internal/pool/pool.go

```diff
@@ -53,18 +62,23 @@ type ConnPool struct {
 	poolTimeout time.Duration
 	idleTimeout time.Duration
 
-	conns     *connList
-	freeConns *connStack
-	stats     PoolStats
+	queue chan struct{}
+
+	connsMu sync.Mutex
+	conns   []*Conn
+
+	freeConnsMu sync.Mutex
+	freeConns   []*Conn
 
-	_closed int32
+	stats PoolStats
 
+	_closed int32 // atomic
 	lastErr atomic.Value
 }
 
 var _ Pooler = (*ConnPool)(nil)
 
-func NewConnPool(dial dialer, poolSize int, poolTimeout, idleTimeout time.Duration) *ConnPool {
+func NewConnPool(dial dialer, poolSize int, poolTimeout, idleTimeout, idleCheckFrequency time.Duration) *ConnPool {
```

Comment:
So pool looks good to me, although I would prefer to abstract `conns` and `freeConns` into their own structs as it makes testing easier and separates trivial and non-trivial logic (again, just my preference).